### PR TITLE
Encode valid empty paths for empty clips

### LINKED ISSUE #644: Corruption issue with clipping
The output rendered by Vello is sometimes corrupted when using clips such as:

```rust
scene.push_layer(Mix::Clip, 1.0, transform, &clip);
scene.pop_layer()
```

This occurs:
- With or without anything being drawn in the clipped layer
- With only small numbers of clips (see below for an example with only 3)
- With no clip nesting (although the example below might nest - not sure)
- With various different blend modes (`Blendmode::default()` and `Mix::Normal` both also exhibit this behaviour)
- With various transform/clip shapes (these do appear to sometimes affect the nature/position of the corruption)

## Runnable example

The `vello-clip-bug` branch of Blitz (https://github.com/DioxusLabs/blitz/tree/vello-clip-bug) reproduces this with `cargo run --release --example google` (also reproduces in debug mode). I think this is the best repro because it happens with only 3 clips (assuming my counting is correct). Vello-related code is almost all in `packages/blitz/src` (and there isn't much else in that dir). The vello setup code is in `renderer.rs`. The vello scene creation code is in `renderer/render.rs`.

There is a suspicion that this could be caused by overlapping clips: https://xi.zulipchat.com/#narrow/stream/197075-gpu/topic/Vello.3A.20.60push_layer.60.20causing.20black.20rectangle

## Screenshots

The issue causes corruption like the following:

Screenshot 2024-07-25 at 20 59 55

It is supposed to look like:

Screenshot 2024-07-26 at 13 06 06

### PR DESCRIPTION
**Pull request by raphlinus**
There was logic to encode a valid empty path when the clip path was empty, but this logic was ineffective because of empty line segment culling. This patch directly encodes an empty path (single zero-length line).

Also includes a doc improvement for `push_layer`.

Fixes #644

### CODE REVIEW

**Review by nicoburns — APPROVED** [review]
I don't have enough context to know whether there is an appropriate fix. But there is nothing obviously wrong with this PR.

File: vello/src/scene.rs [hunk 1]
```diff
@@ -66,6 +66,9 @@ impl Scene {
     /// until the layer is popped.
     ///
     /// **However, the transforms are *not* saved or modified by the layer stack.**
+    ///
+    /// Clip layers (`blend` = [`Mix::Clip`]) should have an alpha value of 1.0.
+    /// For an opacity group with non-unity alpha, specify [`Mix::Normal`].
```
Comment (waywardmonkeys) [inline vello/src/scene.rs:71]: I'd like to reword this to not use the term "non-unity" .. I know what it means, but I don't think all of our readers will.

File: vello/src/scene.rs [hunk 2]
```diff
@@ -80,8 +83,7 @@ impl Scene {
         if !self.encoding.encode_shape(clip, true) {
             // If the layer shape is invalid, encode a valid empty path. This suppresses
             // all drawing until the layer is popped.
-            self.encoding
-                .encode_shape(&Rect::new(0.0, 0.0, 0.0, 0.0), true);
+            self.encoding.encode_empty_shape();
         } else {
```
Comment (waywardmonkeys) [inline vello/src/scene.rs:87]: Should this have a bit for the `bump_estimate` feature to count the 1 segment (empty) path or is that not needed since it is empty?

File: vello_encoding/src/encoding.rs [hunk 3]
```diff
@@ -240,6 +240,12 @@ impl Encoding {
         encoder.finish(true) != 0
     }
 
+    pub fn encode_empty_shape(&mut self) {
```
Comment (waywardmonkeys) [inline vello_encoding/src/encoding.rs:243]: This needs a doc comment.

**Review by waywardmonkeys — COMMENTED** [review]
I feel like the bit about this needing to bypass the path collapsing needs to be captured in a comment somewhere.

I think that we need some infrastructure where we can add code-based tests for things where we validate the encoding. But perhaps not as part of this.  (If we had a way to validate that the encoding is okay, then that would open up to doing some fuzz testing to find ways to generate invalid encodings.)

File: vello/src/scene.rs (hunk 1 above)
Comment (DJMcNab) [inline vello/src/scene.rs:71]: Can we add a warning for this being violated?

**Review by DJMcNab — COMMENTED** [review]
I think it would be worth adding a new screenshot test for this (potentially with a few different variations?). That would also let you test out the new testing workflow.

File: vello/src/scene.rs (hunk 1 above)
Comment (dfrg) [reply vello/src/scene.rs:71]: I believe clip is just an optimization so what about removing it from the public API and having normal+over+opaque encode as a clip?

**Comment by DJMcNab** [comment]
I see that none of the review comments have been addressed. Backed out the merge in case a push has been missed.

**Comment by raphlinus** [comment]
Argh, I was on the bus and missed that there were open review comments. Will address.

File: vello/src/scene.rs (hunk 1 above)
Comment (raphlinus) [reply vello/src/scene.rs:71]: I'm open to adding a warning but don't know what the technology for that is.

And, as discussed in the [Zulip thread](https://xi.zulipchat.com/#narrow/stream/197075-gpu/topic/Vello.3A.20.60push_layer.60.20causing.20black.20rectangle/near/460112832), it's not just an optimization, it has different semantics when doing nontrivial blending. I'll improve the docs a bit. Also, as discussed in that thread, it's likely that support for non-isolated blend groups should become a feature request, but that's out of scope for this PR.

File: vello_encoding/src/encoding.rs (hunk 3 above)
Comment (DJMcNab) [reply vello_encoding/src/encoding.rs:243]: This is still the case

Comment (raphlinus) [reply vello_encoding/src/encoding.rs:248]: Duh, I had forgot to Cmd-S my buffer. Sorry about that; it feels like today is a day I shouldn't be operating heavy machinery.
